app/models/manageiq/providers/redhat/infra_manager/refresher.rb

```diff
@@ -2,18 +2,73 @@ class ManageIQ::Providers::Redhat::InfraManager
   class Refresher < ManageIQ::Providers::BaseManager::Refresher
     include ::EmsRefresh::Refreshers::EmsRefresherMixin
 
-    def parse_legacy_inventory(ems)
-      rhevm = ems.rhevm_inventory
-      raise "Invalid RHEV server ip address." if rhevm.api.nil?
+    def collect_inventory_for_targets(ems, targets)
+      inventory = ems.rhevm_inventory
+      raise "Invalid RHEV server ip address." if inventory.api.nil?
 
-      raw_ems_data = rhevm.refresh
-      return [] if raw_ems_data.blank?
+      # TODO before iterating over targets it would be good to check whether ExtMgmntSystem is part of it
+      # TODO optimize not to fetch the same objects like clusters for multiple targets
 
-      #TODO cleanup with @ems_data
-      ems.api_version = rhevm.service.version_string
+      targets_with_data = targets.collect do |target|
+        _log.info "Filtering inventory for #{target.class} [#{target.name}] id: [#{target.id}]..."
+
+        case target
+          when ExtManagementSystem
+            data,  = Benchmark.realtime_block(:fetch_all) { inventory.refresh }
+
+          when Host
+            methods = {
+                :primary => {
+                    :host                 => target.ems_ref,
+                },
+                :secondary => {
+                    :host                 => [:statistics, :host_nics],
+                }
+            }
+
+            data,  = Benchmark.realtime_block(:fetch_host_data) { data = inventory.targeted_refresh(methods) }
+
+          when VmOrTemplate
+            require 'uri'
+
+            vm = target.ems_ref
+            vm_id = URI(vm).path.split('/').last
+
+            methods = {
+                :primary => {
+                    :cluster     => target.parent_cluster.ems_ref,
+                    :data_center => target.parent_datacenter.ems_ref,
+                    :vm          => vm,
+                    :template    => "/api/templates?search=vm.id=#{vm_id}"
+                },
+                :secondary => {
+                    :vm         => [:disks, :snapshots, :nics],
+                    :template   => [:disks]
+                }
+            }
+            data,  = Benchmark.realtime_block(:fetch_vm_data) { inventory.targeted_refresh(methods) }
+
+        end
+
+        _log.info "Filtering inventory...Complete"
+        [target, data]
+      end
+
+      ems.api_version = inventory.service.version_string
```

Comment:
As a side note, we are eventually going to pull this out, probably into the `EmsRefresherMixin`.  Just FYI...